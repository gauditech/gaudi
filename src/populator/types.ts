type PopulatorModel = {
  model: string;
  repeat: { min: number; max: number } | number;
  data: PopulatorData[];
};

type PopulatorData = {
  field: string;
  type: "constant" | "";
};
