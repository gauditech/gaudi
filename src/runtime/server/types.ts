import { NextFunction, Request, Response } from "express";

export type ServerHttpMethods = "all" | "get" | "post" | "put" | "patch" | "delete";

export type ServerRequestHandler =
  | ((request: Request, response: Response) => void | Promise<void>)
  | ((req: Request, resp: Response, next: NextFunction) => void | Promise<void>);

export type EndpointConfig = {
  path: string;
  method: ServerHttpMethods;
  handlers: ServerRequestHandler[];
};

export type ServerMiddlewareNextFn = (err?: unknown) => void;
