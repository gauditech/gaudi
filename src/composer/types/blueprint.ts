import { Model } from "./model";

export interface Blueprint {
  models: Record<string, Model>;
}
