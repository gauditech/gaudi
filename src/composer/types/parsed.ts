export interface ModelDef {
  name: string;
  fields: FieldDef[];
  references?: ReferenceDef[];
  relations?: RelationDef[];
}

export interface FieldDef {
  name: string;
  type: "string" | "integer" | "datetime";
  nullable?: boolean;
}

export interface ReferenceDef {
  name: string;
  model: string;
}

export interface RelationDef {
  name: string;
  model: string;
}
