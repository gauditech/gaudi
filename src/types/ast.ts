import { WithContext } from "@src/common/error";

export type AST = {
  models: ModelAST[];
};

export type ModelAST = WithContext<{
  name: string;
  body: ModelBodyAST[];
}>;

export type ModelBodyAST = FieldAST | ReferenceAST | RelationAST | QueryAST | ComputedAST;

export type FieldAST = WithContext<{
  kind: "field";
  name: string;
  body: FieldBodyAST[];
}>;

export type FieldBodyAST = WithContext<
  | { kind: "type"; type: string }
  | { kind: "default"; default: LiteralValue }
  | { kind: "tag"; tag: FieldTag }
>;

export type FieldTag = "nullable" | "unique";

export type ReferenceAST = WithContext<{
  kind: "reference";
  name: string;
  body: ReferenceBodyAST[];
}>;

export type ReferenceBodyAST = WithContext<
  { kind: "to"; to: string } | { kind: "tag"; tag: ReferenceTag }
>;

export type ReferenceTag = "nullable" | "unique";

export type RelationAST = WithContext<{
  kind: "relation";
  name: string;
  body: RelationBodyAST[];
}>;

export type RelationBodyAST = WithContext<
  { kind: "from"; from: string } | { kind: "through"; through: string }
>;

export type QueryAST = WithContext<{
  kind: "query";
  name: string;
  body: QueryBodyAST[];
}>;

export type QueryBodyAST = WithContext<
  | { kind: "from"; from: string[] }
  | { kind: "filter"; filter: ExpAST }
  | { kind: "orderBy"; orderings: QueryOrderAST[] }
  | { kind: "limit"; limit: number }
>;

export type QueryOrderAST = WithContext<{ field: string[]; order?: "asc" | "desc" }>;

export type ComputedAST = WithContext<{ kind: "computed"; name: string; exp: ExpAST }>;

export type ExpAST = WithContext<
  | {
      kind: "binary";
      operator: BinaryOperator;
      lhs: ExpAST;
      rhs: ExpAST;
    }
  | { kind: "paren"; exp: ExpAST }
  | { kind: "unary"; operator: UnaryOperator; exp: ExpAST }
  | { kind: "identifier"; identifier: string[] }
  | { kind: "literal"; literal: LiteralValue }
>;

export type Entrypoint = WithContext<{
  name: string;
  body: EntrypointBody[];
}>;

export type EntrypointBody = WithContext<
  | { kind: "target"; identifier: string }
  | { kind: "identify"; identifier: string }
  | { kind: "alias"; identifier: string }
  | { kind: "response"; select: string[] }
  | { kind: "endpoint"; endpoint: Endpoint }
  | { kind: "entrypoint"; entrypoint: Entrypoint }
>;

export type Endpoint = WithContext<{
  type: EndpointType;
  body: EndpointBody;
}>;

export type EndpointType = "list" | "get" | "create" | "update" | "delete";

export type EndpointBody = WithContext<{ kind: "action" }>;

export type LiteralValue = null | boolean | number | string;

export type BinaryOperator =
  | "or"
  | "and"
  | "is not"
  | "is"
  | "not in"
  | "in"
  | "<"
  | "<="
  | ">"
  | ">=";

export type UnaryOperator = "not";
