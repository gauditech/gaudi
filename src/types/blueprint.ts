import { Field, Model } from "./model";

export interface Blueprint {
  models: Record<string, Model>;
  fields: Record<string, Field>;
}
