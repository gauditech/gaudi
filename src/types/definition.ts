export type Definition = {
  models: ModelDef[];
};

export type ModelDef = {
  refKey: string;
  name: string;
  dbname: string;
  fields: FieldDef[];
  references: ReferenceDef[];
  relations: RelationDef[];
  queries: QueryDef[];
};

export type FieldDef = {
  refKey: string;
  modelRefKey: string;
  name: string;
  dbname: string;
  type: "integer" | "text" | "boolean";
  dbtype: "serial" | "integer" | "text" | "boolean";
  primary: boolean;
  unique: boolean;
  nullable: boolean;
};

export type ReferenceDef = {
  refKey: string;
  modelRefKey: string;
  name: string;
  fieldRefKey: string;
  toModelRefKey: string;
  toModelFieldRefKey: string;
  nullable: boolean;
  unique: boolean;
};

export type RelationDef = {
  refKey: string;
  modelRefKey: string;
  name: string;
  fromModel: string;
  fromModelRefKey: string;
  through: string;
  throughRefKey: string;
  nullable: boolean;
  unique: boolean;
};

type QueryDefPathSelect = {
  name: string;
  retType: string;
  nullable: boolean;
  refKey: string;
};

export type QueryDefPath = {
  refKey: string;
  name: string;
  retType: string;
  retCardinality: "one" | "many"; // should be retCard,...
  nullable: boolean;
  joinType: "inner" | "left";
  alias: string;
  bpAlias: string | null;
  joinPaths: QueryDefPath[];
  select: QueryDefPathSelect[];
};

type QueryDefFilter = {
  type: string;
  lhs: string | number | boolean;
  rhs: string | number | boolean;
};

export type QueryDef = {
  refKey: string;
  name: string;
  retType: string;
  retCardinality: "one" | "many";
  nullable: boolean;
  // unique: boolean;
  joinPaths: QueryDefPath[];
  filters: QueryDefFilter[];
};
