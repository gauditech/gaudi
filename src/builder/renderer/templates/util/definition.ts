import { Definition, PathFragment } from "@src/types/definition";

/*
 * Template helper functions for Gaudi definition
 */

export function modelDbName(modelRefKey: string, definition: Definition) {
  return (definition.models || []).find((model) => model.refKey === modelRefKey)?.dbname;
}

export function fieldDbName(modelRefKey: string, fieldRefKey: string, definition: Definition) {
  return (definition.models || [])
    .find((model) => model.refKey === modelRefKey)
    ?.fields.find((field) => field.refKey === fieldRefKey)?.dbname;
}

export function typeToDbType(type: string) {
  if (type === "serial") {
    return "Int";
  } else if (type === "integer") {
    return "Int";
  } else if (type === "text") {
    return "String";
  } else if (type === "boolean") {
    return "Boolean";
  }

  throw "UNHANDLED_GAUDI_TYPE";
}

export function buildEndpointPath(parts: PathFragment[]): string {
  return parts.reduce((accum, part) => {
    const pathPart =
      part.type === "numeric" || part.type === "text" ? `:${part.varname}` : part.value;
    return `${accum}/${pathPart}`;
  }, "");
}
