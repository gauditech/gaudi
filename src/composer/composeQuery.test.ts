import { compileToOldSpec, compose } from "@src/index";
import { CustomOneEndpointDef } from "@src/types/definition";

describe("compose model queries", () => {
  it("nested example without filters", () => {
    const bp = `
    model Org {
      relation repos { from Repo, through org }
      query back_to_org { from repos.org }
    }
    model Repo {
      reference org { to Org }
    }
    `;
    const def = compose(compileToOldSpec(bp));
    expect(def.models[0].queries).toMatchSnapshot();
  });
  it("example with nested filters", () => {
    const bp = `
    model Org {
      relation repos { from Repo, through org }
      query repos_if_one { from repos, filter { org.id is 1 and is_active } }
    }
    model Repo {
      field is_active { type boolean }
      reference org { to Org }
    }
    `;
    const def = compose(compileToOldSpec(bp));

    expect(def.models[0].queries).toMatchSnapshot();
  });

  it("order and limit", () => {
    const bp = `
    model Org {
      relation repos { from Repo, through org }
      query recent_repos {
        from repos,
        order by { id desc },
        limit 5
      }
    }
    model Repo {
      reference org { to Org }
    }
    `;
    const def = compose(compileToOldSpec(bp));
    expect(def.models[0].queries[0]).toMatchSnapshot();
  });
});

describe("compose action queries", () => {
  describe('"fetch" action', () => {
    it("native endpoint", () => {
      const bp = `
      model Org {
        field name { type string }
        field description { type string }
      }
      model Repo {
      }

      api {
        entrypoint Org {
          // test in native endpoint
          update endpoint {
            action {
              update {}
              // target
              fetch as cOrg {
                query { from Org, filter { id is 1 }, select {name} } // TODO: read from ctx - id
              }
              // other model
              fetch as cRepo {
                query { from Repo, filter { id is 1 } } // TODO: read from ctx - id
              }
            }
          }
        }
      }
      `;

      const def = compose(compileToOldSpec(bp));
      expect(
        (def.apis[0].entrypoints[0].endpoints[0] as CustomOneEndpointDef).actions[0]
      ).toMatchSnapshot();
    });
    it("custom endpoint", () => {
      const bp = `
      model Org {
        field name { type string }
        field description { type string }
      }
      model Repo {
      }

      api {
        entrypoint Org {

          // test in custom endpoint
          custom endpoint {
            path "customPath"
            method POST
            cardinality one

            action {
              // target
              fetch as cOrg {
                query { from Org, filter { id is 1 }, select {name} } // TODO: read from ctx - id
              }
              // other model
              fetch as cRepo {
                query { from Repo, filter { id is 1 } } // TODO: read from ctx - id
              }
            }
          }
        }
      }
      `;

      const def = compose(compileToOldSpec(bp));
      expect(
        (def.apis[0].entrypoints[0].endpoints[0] as CustomOneEndpointDef).actions[0]
      ).toMatchSnapshot();
    });
  });
});
