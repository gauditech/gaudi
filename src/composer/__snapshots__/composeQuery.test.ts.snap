// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compose model queries example with nested filters 1`] = `
[
  {
    "ctxModelRefKey": "Org",
    "filter": {
      "kind": "binary",
      "lhs": {
        "kind": "binary",
        "lhs": {
          "kind": "alias",
          "namePath": [
            "org",
            "id",
          ],
        },
        "operator": "is",
        "rhs": {
          "kind": "literal",
          "type": "integer",
          "value": 1,
        },
      },
      "operator": "and",
      "rhs": {
        "kind": "alias",
        "namePath": [
          "is_active",
        ],
      },
    },
    "fromPath": [
      "Org",
      "repos",
    ],
    "joinPaths": [
      {
        "joinPaths": [
          {
            "joinPaths": [],
            "joinType": "inner",
            "kind": "reference",
            "refKey": "Repo.org",
            "retCardinality": "one",
            "retType": "Org",
          },
        ],
        "joinType": "inner",
        "kind": "relation",
        "refKey": "Org.repos",
        "retCardinality": "many",
        "retType": "Repo",
      },
    ],
    "name": "repos_if_one",
    "nullable": false,
    "refKey": "Org.repos_if_one",
    "retCardinality": "many",
    "retType": "Repo",
  },
]
`;

exports[`compose model queries nested example without filters 1`] = `
[
  {
    "ctxModelRefKey": "Org",
    "filter": undefined,
    "fromPath": [
      "Org",
      "repos",
      "org",
    ],
    "joinPaths": [
      {
        "joinPaths": [
          {
            "joinPaths": [],
            "joinType": "inner",
            "kind": "reference",
            "refKey": "Repo.org",
            "retCardinality": "one",
            "retType": "Org",
          },
        ],
        "joinType": "inner",
        "kind": "relation",
        "refKey": "Org.repos",
        "retCardinality": "many",
        "retType": "Repo",
      },
    ],
    "name": "back_to_org",
    "nullable": false,
    "refKey": "Org.back_to_org",
    "retCardinality": "many",
    "retType": "Org",
  },
]
`;
